BUILD

```diff
@@ -13,7 +13,6 @@ cc_library(
         "proxy_wasm_api.h",
         "proxy_wasm_externs.h",
     ],
-    copts = ["-std=c++17"],
```

Comment:
`.bazelrc` flags are not carried over to the dependent projects, so you're effectively removing `-std=c++17` flag when building this as part of other workspaces.

Side-note: With PRs such as this, it's usually a good idea to test it locally with `proxy-wasm-cpp-host` and `envoy` (e.g. using `--override_repository=` flag), and then to open draft PRs pointing to an unmerged commit to verify that it doesn't break anything across those projects.

Reply:
Since the Emscripten update, I am unable to get Abseil (used by protobuf) building without workspace-level copts. AIUI the problem with copts on targets is that those don't propagate to the dependencies. Maybe at issue are transitions or toolchain changes? Extra strange because Emscripten specifies gnu++17 in its toolchain [here](https://github.com/emscripten-core/emsdk/blob/ef2a8e929d5337755e9b1d1e1d4ad859dc694eee/bazel/emscripten_toolchain/toolchain.bzl#L520-L524).

Sounds like you want me to keep digging at this?

Reply:
I don't mind adding workspace-level `cxxopts` to `.bazelrc`, that's not what this comment is about.

I only object to removing them from `cc_library` targets, since you're effectively removing them from builds in the dependent projects (`proxy-wasm-cpp-host`, `envoy`, `istio-proxy`, etc.), which don't inherit the workspace-level `cxxopts` from this workspace.

I suspect most of them build with C++17 anyway, but it's a long chain of dependencies, so it seems like an unnecessary risk.

Reply:
Gotcha thanks, I plan to do two things next:
- figure out if there are any build workarounds using cc_library targets (preferred)
- verify that the repos you mentioned continue to build/test as is

Reply:
> figure out if there are any build workarounds using cc_library targets (preferred)

I struggled with the protobuf v23.4 update for a while. I see at least some of the dependencies (such as the protoc compiler) not being built with emscripten, and not respecting per-target copts either. However, ~everything in protobuf depends on Abseil, and Abseil does require specifying c++14 or higher. In short, I didn't find a good a way to avoid .bazelrc flags with recent versions of protobuf. There is also other fallout from the protobuf update, such as the utf8_range build error and a required update to bazelversion.

I've reverted these changes in the most recent commit. What I propose instead:
- scope this PR more tightly: only update emsdk and disable a zlib warning in the emsdk transition
- [next] work on breaking CppHost --> CppSdk dependency
- [then] try again to update protobuf in CppSdk

> verify that the repos you mentioned continue to build/test as is

Working on this now.

Reply:
https://github.com/envoyproxy/envoy/pull/29118 failed

An error building zlib, which is suspicious because I added `-Wno-deprecated-non-prototype` to fix a warning there. The Envoy build error is wildly different though: `CMAKE_C_COMPILER not set`

```
__________________________
+ /b/f/w/external/cmake-3.23.2-linux-x86_64/bin/cmake -DCMAKE_AR=/b/f/w/external/emsdk/emscripten_toolchain/emar.sh '-DCMAKE_CXX_LINK_EXECUTABLE=/b/f/w/external/emsdk/emscripten_toolchain/emcc_link.sh <FLAGS> <CMAKE_CXX_LINK_FLAGS> <LINK_FLAGS> <OBJECTS> -o <TARGET> <LINK_LIBRARIES>' '-DCMAKE_SHARED_LINKER_FLAGS=--sysroot=/b/f/w/external/emscripten_bin_linux/emscripten/cache/sysroot -fdiagnostics-color -fno-strict-aliasing -funsigned-char -no-canonical-prefixes -s PRINTF_LONG_DOUBLE=1 --oformat=js -O2 -shared -sSTANDALONE_WASM' '-DCMAKE_EXE_LINKER_FLAGS=--sysroot=/b/f/w/external/emscripten_bin_linux/emscripten/cache/sysroot -fdiagnostics-color -fno-strict-aliasing -funsigned-char -no-canonical-prefixes -s PRINTF_LONG_DOUBLE=1 --oformat=js -O2 -sSTANDALONE_WASM' -DCMAKE_CXX_COMPILER_FORCED=on -DCMAKE_C_COMPILER_FORCED=on -DSKIP_BUILD_EXAMPLES=on -DBUILD_SHARED_LIBS=off -DZLIB_COMPAT=on -DZLIB_ENABLE_TESTS=off -DWITH_OPTIM=on -DWITH_SSE4=off -DWITH_NEW_STRATEGIES=off -DUNALIGNED_OK=off -DCMAKE_BUILD_TYPE=Bazel -DCMAKE_INSTALL_PREFIX=/b/f/w/bazel-out/wasm-opt-ST-c90cdbc3697d/bin/external/envoy/bazel/foreign_cc/zlib -DCMAKE_PREFIX_PATH=/b/f/w/bazel-out/wasm-opt-ST-c90cdbc3697d/bin/external/envoy/bazel/foreign_cc/zlib.ext_build_deps -DCMAKE_RANLIB= -DCMAKE_MAKE_PROGRAM=/b/f/w/external/ninja_1.11.0_linux/ninja -G Ninja /b/f/w/external/net_zlib
CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


CMake Error at /b/f/w/external/cmake-3.23.2-linux-x86_64/share/cmake-3.23/Modules/CMakeDetermineCCompiler.cmake:49 (message):
  Could not find compiler set in environment variable CC:

  /b/f/w/external/emsdk/emscripten_toolchain/emcc.sh.
Call Stack (most recent call first):
  CMakeLists.txt:4 (project)


CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
-- Configuring incomplete, errors occurred!
```

Reply:
Emscripten + cmake... I'm way out of my depth here, so I'm calling in the cavalry @walkingeyerobot. For context this build error started happening when upgrading emsdk from v3.1.7 to v3.1.44.

I see CMAKE_C_COMPILER being set here: https://github.com/emscripten-core/emscripten/blob/23949d0989c7039edc3cafef1fe99012d81fffe2/cmake/Modules/Platform/Emscripten.cmake#L90

Has anything changed in how Bazel should use emsdk? It looks like maybe emcmake is expected to add Emscripten.cmake to the toolchain: https://github.com/emscripten-core/emscripten/blob/23949d0989c7039edc3cafef1fe99012d81fffe2/emcmake.py#L34-L36

Here's the relevant/failing target (though it builds fine for me locally...):
https://github.com/envoyproxy/envoy/blob/d191b2a583d693c29243b4722da3508eabe2d71c/bazel/foreign_cc/BUILD#L519

Reply:
Yup, here you go:
1. Check out https://github.com/envoyproxy/envoy/pull/29118.
2. If you want to reduce the scope, you can revert changes to `proxy_wasm_cpp_sdk` in `bazel/repository_locations.bzl`. The repro case only requires updating emsdk v3.1.7 -> v3.1.44.
3. Build a `envoy_wasm_cc_binary`, such as:
    `$ bazelisk build -s //test/extensions/filters/http/wasm/test_data:test_cpp.wasm`

Reply:
Emscripten is definitely being asked to build zlib as a cmake target under Envoy, and looking at blame this hasn't changed in 2-4 years: https://github.com/envoyproxy/envoy/blob/ee594d24dbf4743212cd6e23bd8cde7105a5b6f8/bazel/repositories.bzl#L545-L562

I think this dependency happens via protobuf: https://github.com/protocolbuffers/protobuf/blob/f881c8fe15b8c9aa3be997103eee5f2bea02984d/src/google/protobuf/io/BUILD.bazel#L145

```
$ bazelisk cquery 'deps(//test/extensions/filters/http/wasm/test_data:test_cpp.wasm)' | grep zlib | grep foreign
@envoy//bazel/foreign_cc:zlib (d073d29)

$ bazelisk config d073d29 | grep emscripten
  compiler: emscripten
  crosstool_top: @emsdk//emscripten_toolchain:everything
  custom_malloc: @emsdk//emscripten_toolchain:malloc
```

Reply:
I have a lead here: using rules_foreign_cc we can likely register a wasm cmake toolchain pointing to emcmake, in emsdk.
https://github.com/bazelbuild/rules_foreign_cc/blob/main/toolchains/BUILD.bazel

Reply:
Bleh, I can't get my emcmake toolchain (https://github.com/martijneken/emsdk/commit/a871cd0e4f4561ec8cd222682102b1aa87b81df5) to be resolved properly. Bazel uses `@cmake-3.23.2-linux-x86_64` instead.

```
$ bazelisk build --toolchain_resolution_debug=zlib //test/extensions/filters/http/wasm/test_data:test_cpp.wasm
...
INFO: ToolchainResolution:     Type @rules_foreign_cc//toolchains:cmake_toolchain: execution platform @local_config_platform//:host: Rejected toolchain @cmake-3.23.2-linux-aarch64//:cmake_tool; mismatching values: arm64
INFO: ToolchainResolution:   Type @rules_foreign_cc//toolchains:cmake_toolchain: target platform @emsdk//:platform_wasm: execution @local_config_platform//:host: Selected toolchain @cmake-3.23.2-linux-x86_64//:cmake_tool
INFO: ToolchainResolution:     Type @rules_foreign_cc//toolchains:ninja_toolchain: target platform @emsdk//:platform_wasm: execution platform @local_config_platform//:host: Skipping toolchain @rules_foreign_cc//toolchains:preinstalled_ninja; execution platform already has selected toolchain
INFO: ToolchainResolution:     Type @rules_foreign_cc//toolchains:cmake_toolchain: target platform @emsdk//:platform_wasm: execution platform @local_config_platform//:host: Skipping toolchain @cmake-3.23.2-macos-universal//:cmake_tool; execution platform already has selected toolchain
INFO: ToolchainResolution:     Type @rules_foreign_cc//toolchains:cmake_toolchain: target platform @emsdk//:platform_wasm: execution platform @local_config_platform//:host: Skipping toolchain @cmake-3.23.2-windows-i386//:cmake_tool; execution platform already has selected toolchain
INFO: ToolchainResolution:     Type @rules_foreign_cc//toolchains:cmake_toolchain: target platform @emsdk//:platform_wasm: execution platform @local_config_platform//:host: Skipping toolchain @cmake-3.23.2-windows-x86_64//:cmake_tool; execution platform already has selected toolchain
INFO: ToolchainResolution:     Type @rules_foreign_cc//toolchains:cmake_toolchain: target platform @emsdk//:platform_wasm: execution platform @local_config_platform//:host: Skipping toolchain @rules_foreign_cc//toolchains:built_cmake; execution platform already has selected toolchain
INFO: ToolchainResolution:     Type @rules_foreign_cc//toolchains:cmake_toolchain: target platform @emsdk//:platform_wasm: execution platform @local_config_platform//:host: Skipping toolchain @rules_foreign_cc//toolchains:preinstalled_cmake; execution platform already has selected toolchain
INFO: ToolchainResolution:     Type @rules_foreign_cc//toolchains:cmake_toolchain: target platform @emsdk//:platform_wasm: execution platform @local_config_platform//:host: Skipping toolchain @emsdk//emscripten_toolchain:cmake-compiler-wasm; execution platform already has selected toolchain

INFO: ToolchainResolution: Target platform @emsdk//:platform_wasm: Selected execution platform @local_config_platform//:host, 
type @rules_foreign_cc//toolchains:make_toolchain -> toolchain @rules_foreign_cc//toolchains:built_make, 
type @bazel_tools//tools/cpp:toolchain_type -> toolchain @rules_rust//rust/private/dummy_cc_toolchain:dummy_cc_wasm32_toolchain_cc, 
type @rules_foreign_cc//toolchains:m4_toolchain -> toolchain @rules_foreign_cc//toolchains:preinstalled_m4, 
type @rules_foreign_cc//toolchains:cmake_toolchain -> toolchain @cmake-3.23.2-linux-x86_64//:cmake_tool, 
type @rules_foreign_cc//foreign_cc/private/framework:shell_toolchain -> toolchain @rules_foreign_cc_framework_toolchain_linux//:commands, 
type @rules_foreign_cc//toolchains:ninja_toolchain -> toolchain @ninja_1.11.0_linux//:ninja_tool, 
type @rules_foreign_cc//toolchains:pkgconfig_toolchain -> toolchain @rules_foreign_cc//toolchains:preinstalled_pkgconfig
```

Reply:
I'm pretty sure the above are Emscripten issues, so I've filed https://github.com/emscripten-core/emsdk/issues/1273

Reply:
Thanks much! Seemingly the last remaining issue is with the emsdk patch file that removes npm targets:
https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/master/bazel/emsdk.patch

I'll have to chase down why that was added and why it's breaking now.

Reply:
The NPM targets were removed, because those downloads were failing a lot (like, completely breaking Envoy's CI) and they were done outside of Bazel's caching system, so they were not cached, and coincidentally those targets were already included in emsdk's tarballs, so removing those fixed the CI at the time.